### ISSUE TITLE
All zarr gcs datastores are broken on pangeo.pydata.org

### ISSUE THREAD

**Issue by rabernat** [pangeo-data/pangeo#176]
Something has changed with the configuration of the cluster. Now I can't read any zarr stores.

In the original xarray-data.ipynb example, we do this:
```python
fs = gcsfs.GCSFileSystem(project='pangeo-181919', token='anon', access='read_only')
gcsmap = gcsfs.mapping.GCSMap('pangeo-data/newman-met-ensemble',
                              gcs=fs, check=False, create=False)
```

That gives the error
```
_call exception: Anonymous caller does not have storage.buckets.list access to project 464800473488.
Traceback (most recent call last):
  File "/home/jovyan/.local/lib/python3.6/site-packages/gcsfs/core.py", line 432, in _call
    validate_response(r, path)
  File "/home/jovyan/.local/lib/python3.6/site-packages/gcsfs/core.py", line 162, in validate_response
    raise HtmlError(error)
gcsfs.utils.HtmlError: Anonymous caller does not have storage.buckets.list access to project 464800473488.
---------------------------------------------------------------------------
HtmlError                                 Traceback (most recent call last)
<ipython-input-3-5e460340597b> in <module>()
     10 fs = gcsfs.GCSFileSystem(project='pangeo-181919', token='anon', access='read_only')
     11 gcsmap = gcsfs.mapping.GCSMap('pangeo-data/newman-met-ensemble',
---> 12                               gcs=fs, check=False, create=False)
     13 
     14 ds = xr.open_zarr(gcsmap)

~/.local/lib/python3.6/site-packages/gcsfs/mapping.py in __init__(self, root, gcs, check, create)
     41             if create:
     42                 self.gcs.mkdir(bucket)
---> 43             elif not self.gcs.exists(bucket):
     44                 raise ValueError("Bucket %s does not exist."
     45                         " Create bucket with the ``create=True`` keyword" %

<decorator-gen-140> in exists(self, path)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in exists(self, path)
    768                 return bool(self.info(path))
    769             else:
--> 770                 if bucket in self.buckets:
    771                     return True
    772                 else:

~/.local/lib/python3.6/site-packages/gcsfs/core.py in buckets(self)
    449     def buckets(self):
    450         """Return list of available project buckets."""
--> 451         return [b["name"] for b in self._list_buckets()["items"]]
    452 
    453     @classmethod

<decorator-gen-131> in _list_buckets(self)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _list_buckets(self)
    568         items = []
    569         page = self._call(
--> 570             'get', 'b/', project=self.project
    571         )
    572 

<decorator-gen-126> in _call(self, method, path, *args, **kwargs)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _call(self, method, path, *args, **kwargs)
    439                     # retry
    440                     continue
--> 441                 raise e
    442         try:
    443             out = r.json()

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _call(self, method, path, *args, **kwargs)
    430                 time.sleep(2**retry - 1)
    431                 r = meth(self.base + path, params=kwargs, json=json)
--> 432                 validate_response(r, path)
    433                 break
    434             except (HtmlError, RequestException, GoogleAuthError) as e:

~/.local/lib/python3.6/site-packages/gcsfs/core.py in validate_response(r, path)
    160             raise ValueError("Bad Request: %s\n%s" % (path, msg))
    161         elif error:
--> 162             raise HtmlError(error)
    163         else:
    164             raise RuntimeError(m)

HtmlError: Anonymous caller does not have storage.buckets.list access to project 464800473488.
```

If I remove the authentication stuff
```python
gcsmap = gcsfs.mapping.GCSMap('pangeo-data/newman-met-ensemble')
```

I get this error
```
_call exception: No such object: pangeo-data/newman-met-ensemble/.zarray
Traceback (most recent call last):
  File "/home/jovyan/.local/lib/python3.6/site-packages/gcsfs/core.py", line 432, in _call
    validate_response(r, path)
  File "/home/jovyan/.local/lib/python3.6/site-packages/gcsfs/core.py", line 162, in validate_response
    raise HtmlError(error)
gcsfs.utils.HtmlError: No such object: pangeo-data/newman-met-ensemble/.zarray
---------------------------------------------------------------------------
HtmlError                                 Traceback (most recent call last)
<ipython-input-2-e81e004b2748> in <module>()
     12                               #, gcs=fs, check=False, create=False)
     13 
---> 14 ds = xr.open_zarr(gcsmap)
     15 
     16 # Print dataset

/opt/conda/lib/python3.6/site-packages/xarray/backends/zarr.py in open_zarr(store, group, synchronizer, auto_chunk, decode_cf, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables)
    457     zarr_store = ZarrStore.open_group(store, mode=mode,
    458                                       synchronizer=synchronizer,
--> 459                                       group=group)
    460     ds = maybe_decode_store(zarr_store)
    461 

/opt/conda/lib/python3.6/site-packages/xarray/backends/zarr.py in open_group(cls, store, mode, synchronizer, group, writer)
    258                                       "#installation" % min_zarr)
    259         zarr_group = zarr.open_group(store=store, mode=mode,
--> 260                                      synchronizer=synchronizer, path=group)
    261         return cls(zarr_group, writer=writer)
    262 

/opt/conda/lib/python3.6/site-packages/zarr/hierarchy.py in open_group(store, mode, cache_attrs, synchronizer, path)
   1108 
   1109     if mode in ['r', 'r+']:
-> 1110         if contains_array(store, path=path):
   1111             err_contains_array(path)
   1112         elif not contains_group(store, path=path):

/opt/conda/lib/python3.6/site-packages/zarr/storage.py in contains_array(store, path)
     70     prefix = _path_to_prefix(path)
     71     key = prefix + array_meta_key
---> 72     return key in store
     73 
     74 

~/.local/lib/python3.6/site-packages/gcsfs/mapping.py in __contains__(self, key)
     86 
     87     def __contains__(self, key):
---> 88         return self.gcs.exists(self._key_to_str(key))
     89 
     90     def __len__(self):

<decorator-gen-140> in exists(self, path)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in exists(self, path)
    766         try:
    767             if key:
--> 768                 return bool(self.info(path))
    769             else:
    770                 if bucket in self.buckets:

<decorator-gen-141> in info(self, path)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in info(self, path)
    800 
    801         try:
--> 802             return self._get_object(path)
    803         except FileNotFoundError:
    804             logger.debug("info FileNotFound at path: %s", path)

<decorator-gen-127> in _get_object(self, path)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _get_object(self, path)
    490 
    491         result = self._process_object(bucket, self._call('get', 'b/{}/o/{}',
--> 492                                                          bucket, key))
    493 
    494         return result

<decorator-gen-126> in _call(self, method, path, *args, **kwargs)

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _call(self, method, path, *args, **kwargs)
    439                     # retry
    440                     continue
--> 441                 raise e
    442         try:
    443             out = r.json()

~/.local/lib/python3.6/site-packages/gcsfs/core.py in _call(self, method, path, *args, **kwargs)
    430                 time.sleep(2**retry - 1)
    431                 r = meth(self.base + path, params=kwargs, json=json)
--> 432                 validate_response(r, path)
    433                 break
    434             except (HtmlError, RequestException, GoogleAuthError) as e:

~/.local/lib/python3.6/site-packages/gcsfs/core.py in validate_response(r, path)
    160             raise ValueError("Bad Request: %s\n%s" % (path, msg))
    161         elif error:
--> 162             raise HtmlError(error)
    163         else:
    164             raise RuntimeError(m)

HtmlError: No such object: pangeo-data/newman-met-ensemble/.zarray
```

This has come up in #174.

**Comment by rabernat** [comment]
Can anyone else verify this? Or is it just my user account?

**Comment by martindurant** [comment]
`elif not self.gcs.exists(bucket)` , this was fixed in https://github.com/dask/gcsfs/pull/88

**Comment by mrocklin** [comment]
I was planning to push an update to the cluster at some point today.  I'll
bump that in my priority list and handle it soon.  I'll point to gcsfs
master and we can see if that resolves things

**Comment by rabernat** [comment]
Can you help me understand what caused this issue to appear suddenly? How did a new gcsfs version make it onto pangeo.pydata.org without someone rebuilding the docker images?

**Comment by martindurant** [comment]
Given the error, I'm surprised it worked before - perhaps the user permissions really did change on the google side. We did see this exception somewhere along the road, though, which is why the PR above was made.

**Comment by mrocklin** [comment]
No, I can not help with that.  I don't know what would have caused that.

**Comment by rabernat** [comment]
> We did see this exception somewhere along the road,

Which exception does dask/gcsfs#88 fix? There are two distinct errors reported above.

**Comment by martindurant** [comment]
It fixes the upper error, where the anonymous user cannot list buckets, but this is not actually necessary to access a mapping.

**Comment by rabernat** [comment]
The lower error involves a call to `self.gcs.exists(self._key_to_str(key))`. Presumably that is related to the same underlying issue I hope.

**Comment by martindurant** [comment]
The bottom error refers specifically to a `.zarray` key, which the code leading up to there does not mention at all, so I don't know where this comes from. In fact, that directory seems to contain `.zgroup` and `.zattrs`. Maybe, indeed, it would be best to delete the directory completely before attempting to write anything (this should be doable with gcsfs without dropping to the gsutil CLI).

**Comment by rabernat** [comment]
To be clear, all of these datasets worked perfectly yesterday (and for several months before); nothing with their internal structure has been changed recently.

My interpretation of the error is that zarr is checking for the existence of the `.zarray` object, causing gcsfs to call `exists`, which is raising some unexpected exception. `.zarray` is *not supposed* to be at that location and indeed it is not there. But zarr can't confirm that, because the call to `gcsfs.exists` errors.

**Comment by mrocklin** [comment]
Given that gcsfs hasn't changed recently perhaps it is not a problem with
gcsfs?

**Comment by rabernat** [comment]
Can anyone reproduce?

**Comment by kaipak** [comment]
I get a similar error

```
_call exception: No such object: pangeo-data/kai-llc4320-vertical-fluxes/llc_4320_agulhas.0000197712/.zarray
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/gcsfs/core.py", line 432, in _call
    validate_response(r, path)
  File "/opt/conda/lib/python3.6/site-packages/gcsfs/core.py", line 162, in validate_response
    raise HtmlError(error)
gcsfs.utils.HtmlError: No such object: pangeo-data/kai-llc4320-vertical-fluxes/llc_4320_agulhas.0000197712/.zarray

---------------------------------------------------------------------------
HtmlError                                 Traceback (most recent call last)
<ipython-input-11-71030b4d49f9> in <module>()
      1 # Open up data and merge
----> 2 ds = xr.open_zarr(llc_data)
      3 ds = ds.set_coords(['iter', 'face'])
      4 grid = xr.open_zarr(llc_grid, auto_chunk=False)
      5 ds = xr.merge([grid, ds])

/opt/conda/lib/python3.6/site-packages/xarray/backends/zarr.py in open_zarr(store, group, synchronizer, auto_chunk, decode_cf, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables)
    457     zarr_store = ZarrStore.open_group(store, mode=mode,
    458                                       synchronizer=synchronizer,
--> 459                                       group=group)
    460     ds = maybe_decode_store(zarr_store)
    461 

/opt/conda/lib/python3.6/site-packages/xarray/backends/zarr.py in open_group(cls, store, mode, synchronizer, group, writer)
    258                                       "#installation" % min_zarr)
    259         zarr_group = zarr.open_group(store=store, mode=mode,
--> 260                                      synchronizer=synchronizer, path=group)
    261         return cls(zarr_group, writer=writer)
    262 

/opt/conda/lib/python3.6/site-packages/zarr/hierarchy.py in open_group(store, mode, cache_attrs, synchronizer, path)
   1108 
   1109     if mode in ['r', 'r+']:
-> 1110         if contains_array(store, path=path):
   1111             err_contains_array(path)
   1112         elif not contains_group(store, path=path):

/opt/conda/lib/python3.6/site-packages/zarr/storage.py in contains_array(store, path)
     70     prefix = _path_to_prefix(path)
     71     key = prefix + array_meta_key
---> 72     return key in store
     73 
     74 

/opt/conda/lib/python3.6/site-packages/gcsfs/mapping.py in __contains__(self, key)
     85 
     86     def __contains__(self, key):
---> 87         return self.gcs.exists(self._key_to_str(key))
     88 
     89     def __len__(self):

<decorator-gen-140> in exists(self, path)

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in exists(self, path)
    766         try:
    767             if key:
--> 768                 return bool(self.info(path))
    769             else:
    770                 if bucket in self.buckets:

<decorator-gen-141> in info(self, path)

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in info(self, path)
    800 
    801         try:
--> 802             return self._get_object(path)
    803         except FileNotFoundError:
    804             logger.debug("info FileNotFound at path: %s", path)

<decorator-gen-127> in _get_object(self, path)

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _get_object(self, path)
    490 
    491         result = self._process_object(bucket, self._call('get', 'b/{}/o/{}',
--> 492                                                          bucket, key))
    493 
    494         return result

<decorator-gen-126> in _call(self, method, path, *args, **kwargs)

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _tracemethod(f, self, *args, **kwargs)
     49         logger.log(logging.DEBUG - 1, tb_io.getvalue())
     50 
---> 51     return f(self, *args, **kwargs)
     52 
     53 

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _call(self, method, path, *args, **kwargs)
    439                     # retry
    440                     continue
--> 441                 raise e
    442         try:
    443             out = r.json()

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in _call(self, method, path, *args, **kwargs)
    430                 time.sleep(2**retry - 1)
    431                 r = meth(self.base + path, params=kwargs, json=json)
--> 432                 validate_response(r, path)
    433                 break
    434             except (HtmlError, RequestException, GoogleAuthError) as e:

/opt/conda/lib/python3.6/site-packages/gcsfs/core.py in validate_response(r, path)
    160             raise ValueError("Bad Request: %s\n%s" % (path, msg))
    161         elif error:
--> 162             raise HtmlError(error)
    163         else:
    164             raise RuntimeError(m)

HtmlError: No such object: pangeo-data/kai-llc4320-vertical-fluxes/llc_4320_agulhas.0000197712/.zarray
```

**Comment by jhamman** [comment]
Not sure if this is related to #172 but, @tjcrone did you push the rbac stuff to gce yesterday?

**Comment by martindurant** [comment]
When recently creating a GKE cluster, I got the following warning
```
WARNING: Starting in Kubernetes v1.10, new clusters will no longer get compute-rw and storage-ro scopes added to what is specified in --scopes (though the latter will remain included in the default --scopes). To use these scopes, add them explicitly to --scopes. To use the new behavior, set container/new_scopes_behavior property to true.
```
Might kubernetes on the cluster have been updated?

**Comment by mrocklin** [comment]
I'm seeing 1.9.3

**Comment by martindurant** [comment]
OK, I believe that the issue is, that the Google error message has changed it's format. Previously it had the words "Not Found" in the message, and this was used to produce a FileNotFound exception. Now it says "No such object". This is annoying, but can be fixed.

**Comment by tjcrone** [comment]
@jhamman I did not merge, and I do not believe anyone else has pulled this yet. I am typically very averse to merging my own commits into master. I think we are waiting for more testing and review before we merge https://github.com/pangeo-data/pangeo/pull/172. Looks like @martindurant may have found the issue.

**Comment by rabernat** [comment]
> I did not merge, and I do not believe anyone else has pulled this yet. I am typically very averse to merging my own commits into master.

Just so we are all on the same page: merging things into master on this repo does not cause them to be deployed to pangeo.pydata.org. The redeployment is still a manual process. We would eventually like to automate it eventually (see #131).

@martindurant - sounds like you have identified the problem indeed. Very annoying for google to change its API without warning! I imagine that other people besides us have been affected by this.

**Comment by mrocklin** [comment]
I believe that this is now resolved in #177 , which is live on the cluster
